system/lib/wasmfs/file_table.cpp

```diff
@@ -0,0 +1,151 @@
+    uint8_t current = buf[j];
+    if (current == '\0' || current == '\n') {
+      fd_write_buffer.push_back('\0'); // for null-terminated C strings
+      console_write(&fd_write_buffer[0]);
+      fd_write_buffer.clear();
+    } else {
+      fd_write_buffer.push_back(current);
+    }
+  }
+  return __WASI_ERRNO_SUCCESS;
+}
+
+class StdinFile : public File {
+
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return __WASI_ERRNO_INVAL;
+  }
+
+  __wasi_errno_t read(const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) override {
+    emscripten_console_log("StdinFile::read() has not been implemented yet.");
+    return __WASI_ERRNO_INVAL;
+  };
+
+public:
+  File::Handle get() { return Handle(*this); }
+};
+
+class StdoutFile : public File {
+  static std::vector<char> writeBuffer;
+
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return writeStdBuffer(buf, len, &emscripten_console_log, writeBuffer);
+  }
+
+  __wasi_errno_t read(const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) override {
+    emscripten_console_log("StdoutFile::read() has not been implemented yet.");
+    return __WASI_ERRNO_INVAL;
+  };
+
+public:
+  File::Handle get() { return Handle(*this); }
+};
+
+class StderrFile : public File {
+  static std::vector<char> writeBuffer;
+
+  // TODO: May not want to proxy stderr (fd == 2) to the main thread.
+  // This will not show in HTML - a console.warn in a worker is sufficient.
+  // This would be a change from the current FS.
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return writeStdBuffer(buf, len, &emscripten_console_error, writeBuffer);
+  }
+
+  __wasi_errno_t read(const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) override {
+    emscripten_console_log("StderrFile::read() has not been implemented yet.");
+    return __WASI_ERRNO_INVAL;
+  };
+
+public:
+  File::Handle get() { return Handle(*this); }
+};
+
+std::vector<char> StdoutFile::writeBuffer;
+std::vector<char> StderrFile::writeBuffer;
+
+std::shared_ptr<File>& OpenFileState::Handle::getFile() { return openFileState.file; }
+
+FileTable::Handle FileTable::get() {
+  static FileTable fileTable;
+  return FileTable::Handle(fileTable);
+}
+
+FileTable::FileTable() {
+  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdinFile>()));
+  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdoutFile>()));
+  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StderrFile>()));
+}
+
+__wasi_fd_t FileTable::Handle::add(std::shared_ptr<OpenFileState> ptr) {
+  for (__wasi_fd_t i = 0; i < fileTable.entries.size(); i++) {
+    if (!fileTable.entries[i]) {
+      // Free open file entry.
+      fileTable.entries[i] = ptr;
+      return i;
+    }
+  }
+
+  // Could not find an empty open file table entry.
+  fileTable.entries.push_back(ptr);
+
+  return fileTable.entries.size() - 1;
```

Comment:
Then `Entry` API can simplify this code as well:
```suggestion
  for (__wasi_fd_t i = 0; ; i++) {
    if (!fileTable[i]) {
      // Free open file entry.
      fileTable[i] = ptr;
      return i;
    }
  }
```
This might even be simple enough that we can move it into the caller (`dup`) and get rid of this method.